sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,413 @@
+ * - A second traversal checks if a [[ScalarSubqueryReference]] is pointing to a merged subquery
+ *   plan or not and either keeps the reference or restores the original [[ScalarSubquery]].
+ *   If there are [[ScalarSubqueryReference]] nodes remained a [[CommonScalarSubqueries]] root node
+ *   is added to the plan with the referenced scalar subqueries.
+ * - [[PlanSubqueries]] or [[PlanAdaptiveSubqueries]] rule does the physical planning of scalar
+ *   subqueries including the ones under [[CommonScalarSubqueriesExec]] node and replaces
+ *   each [[ScalarSubqueryReference]] to their referenced physical plan in
+ *   `GetStructField(ScalarSubquery(merged plan with CreateNamedStruct() header))` form.
+ *   It is important that references pointing to the same merged subquery are replaced to the same
+ *   planned instance to make sure that each merged subquery runs only once (even without a wrapping
+ *   [[ReuseSubquery]] node).
+ *   Finally, the [[CommonScalarSubqueriesExec]] node is removed from the physical plan.
+ * - The [[ReuseExchangeAndSubquery]] rule wraps the second, third, ... instances of the same
+ *   subquery into a [[ReuseSubquery]] node, but this just a cosmetic change in the plan.
+ *
+ * Eg. the following query:
+ *
+ * SELECT
+ *   (SELECT avg(a) FROM t GROUP BY b),
+ *   (SELECT sum(b) FROM t GROUP BY b)
+ *
+ * is optimized from:
+ *
+ * Project [scalar-subquery#231 [] AS scalarsubquery()#241,
+ *          scalar-subquery#232 [] AS scalarsubquery()#242L]
+ * :  :- Aggregate [b#234], [avg(a#233) AS avg(a)#236]
+ * :  :  +- Relation default.t[a#233,b#234] parquet
+ * :  +- Aggregate [b#240], [sum(b#240) AS sum(b)#238L]
+ * :     +- Project [b#240]
+ * :        +- Relation default.t[a#239,b#240] parquet
+ * +- OneRowRelation
+ *
+ * to:
+ *
+ * CommonScalarSubqueries [scalar-subquery#250 []]
+ * :  +- Project [named_struct(avg(a), avg(a)#236, sum(b), sum(b)#238L) AS mergedValue#249]
+ * :     +- Aggregate [b#234], [avg(a#233) AS avg(a)#236, sum(b#234) AS sum(b)#238L]
+ * :        +- Project [a#233, b#234]
+ * :           +- Relation default.t[a#233,b#234] parquet
+ * +- Project [scalarsubqueryreference(0, 0, DoubleType, 231) AS scalarsubquery()#241,
+ *             scalarsubqueryreference(0, 1, LongType, 232) AS scalarsubquery()#242L]
+ *    +- OneRowRelation
+ */
+object MergeScalarSubqueries extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (conf.subqueryReuseEnabled) {
+      plan match {
+        case Subquery(_: CommonScalarSubqueries, _) => plan
+        case s: Subquery => s.copy(child = extractCommonScalarSubqueries(s.child))
+        case _: CommonScalarSubqueries => plan
+        case _ => extractCommonScalarSubqueries(plan)
+      }
+    } else {
+      plan
+    }
+  }
+
+  private def extractCommonScalarSubqueries(plan: LogicalPlan) = {
+    // Plan of subqueries and a flag is the plan is merged
+    val cache = ListBuffer.empty[(Project, Boolean)]
```

Comment:
more comments regarding the cache? maybe the the flag should be changed to int to record how many times it is referenced.

Reply:
is it worthwhile to alias Project to some different type name that hint this is the merged header of scalar query ?

Reply:
> is it worthwhile to alias Project
Thanks, I will change it soon.

> maybe the the flag should be changed to int to record how many times it is referenced.
But why do we need that? What we need to know about an item of the cache is if it is a result of merge(s) or an original query.

Reply:
> > is it worthwhile to alias Project to some different type name that hint this is the merged header of scalar query ?
> 
> Thanks, I will change it soon.

Fixed in https://github.com/apache/spark/pull/32298/commits/d86d2c48a3e6244fc091ae09cb9377ade98f66b0.